crates/moon/src/rr_build/mod.rs

```diff
@@ -840,3 +842,70 @@ pub fn execute_build_partial(
 
     Ok(stats)
 }
+
+/// The callback function that catches diagnostics output. The output is
+/// expected to have json format. The captured diagnostics will be processed
+/// later.
+fn capture_diagnostics_callback(catcher: Arc<Mutex<ResultCatcher>>) -> impl Fn(&str) {
+    move |output: &str| {
+        output
+            .split('\n')
+            .filter(|it| !it.is_empty())
+            .for_each(|content| {
+                catcher.lock().unwrap().append_content(content, None);
```

Comment:
Alternatives:
- Consider only accumulating diagnostic instances, and directly output the others
- Consider printing diagnostics on first sighting, and then add it to the map, so it doesn't get printed twice. (I see you're using a BTreeSet to collect, so this interferes with ordering)